### PR TITLE
 Set collection if there is not any collection before

### LINKED ISSUE #32229: [Issue]  Set collection if there is not any collection before
This issue is automatically created based on existing pull request: magento/magento2#27313:  Set collection if there is not any collection before

---------

### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/341

### Fixed Issues (if relevant)
1.https://github.com/magento/partners-magento2ee/issues/64

### Manual testing scenarios (*)
Login to admin -> Go to catalog -> Category
Under products in category section
Filterable section is available and filtrable

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by Nazar65**
### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/516

### Fixed Issues (if relevant)
1.https://github.com/magento/partners-magento2ee/issues/64

### Manual testing scenarios (*)
Login to admin -> Go to catalog -> Category
Under products in category section
Filterable section is available and filtrable

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#32229:  Set collection if there is not any collection before

### CODE REVIEW

**Comment by engcom-Golf** [comment]
@magento run all tests

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Please fix the failing Unit Tests and Integration Tests

**Comment by engcom-Golf** [comment]
> Please fix the failing Unit Tests and Integration Tests
@lbajsarowicz already in progress

**Comment by engcom-Golf** [comment]
@lbajsarowicz :heavy_check_mark:

File: app/code/Magento/Catalog/Block/Adminhtml/Category/Tab/Product.php [hunk 1]
```diff
@@ -110,37 +117,41 @@ protected function _addColumnFilterToCollection($column)
...
+            )->addAttributeToSelect(
+                'price'
+            )->joinField(
+                'position',
+                'catalog_category_product',
+                'position',
+                'product_id=entity_id',
+                'category_id=' . (int)$this->getRequest()->getParam('id', 0),
+                'left'
+            );
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Block/Adminhtml/Category/Tab/Product.php]: I know that's about the preferences, but I personally hate these line-breaks.
```suggestion
            $collection = $this->_productFactory->create()->getCollection()
                ->addAttributeToSelect('name')
                ->addAttributeToSelect('sku')
                ->addAttributeToSelect('visibility')
                ->addAttributeToSelect('status')
                ->addAttributeToSelect('price')
                ->joinField(
                    'position',
                    'catalog_category_product',
                    'position',
                    'product_id=entity_id',
                    ['category_id' => (int)$this->getRequest()->getParam('id')],
                    'left'
                );
```

**Review by lbajsarowicz — DISMISSED** [review]
I recommend using different formatting for the filters block. Also instead of coalescing - use array for filters.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7149](https://jira.corp.magento.com/browse/ENGCOM-7149) has been created to process this Pull Request

**Review by sidolov — CHANGES_REQUESTED** [review]
@Nazar65 please, apply review changes and cover them with automated tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@engcom-Foxtrot Without code changes you don't need to run all the tests, but only the failed ones.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        <annotations>
+            <stories value="Product"/>
+            <features value="Catalog"/>
+            <title value="Admin category product filters test"/>
+            <description value="Check if Product Grid filters work correctly on Category page"/>
+            <severity value="MAJOR"/>
+            <group value="catalog"/>
+        </annotations>
+        <before>
+            <!-- Create Category -->
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml:21]: You don't need to provide comments, as these are displayed just like other steps.
It's enough that `stepKey` for the actual step is descriptive.

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            <createData entity="SimpleSubCategory" stepKey="createCategory"/>
+
+            <!-- Create Products -->
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml:24]: ↑ Same.

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            <createData entity="SimpleProduct" stepKey="createSimpleProductFirst">
+                <requiredEntity createDataKey="createCategory"/>
+            </createData>
+            <createData entity="SimpleProduct" stepKey="createSimpleProductSecond">
+                <requiredEntity createDataKey="createCategory"/>
+            </createData>
+
+            <!--Login as admin -->
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml:32]: ↑ Same.

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            <deleteData createDataKey="createCategory" stepKey="deleteCategory"/>
+            <actionGroup ref="AdminLogoutActionGroup" stepKey="logout"/>
+        </after>
+
+        <!-- Open Category -->
+        <actionGroup ref="GoToAdminCategoryPageByIdActionGroup" stepKey="openCategory">
+            <argument name="id" value="$createCategory.id$"/>
+        </actionGroup>
+
+        <!-- Check if Filterable Section is available -->
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml:47]: Same as above.

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        <actionGroup ref="AdminCategoryPageOpenProductsInCategorySectionActionGroup"
+                     stepKey="openProductsInCategorySection"/>
+
+        <!-- Apply Sku Filter -->
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml:51]: Same as above.

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        <actionGroup ref="AdminCategoryFilterProductBySkuActionGroup" stepKey="applySkuFilter">
+            <argument name="sku" value="$createSimpleProductFirst.sku$"/>
+        </actionGroup>
+
+        <!-- Check Sku Filter correctly applied -->
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml:56]: Same as above.

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        <actionGroup ref="AdminCheckProductPositionInCategoryProductsGridActionGroup" stepKey="seeFirstProductFiltered">
+            <argument name="position" value="1"/>
+            <argument name="productName" value="$createSimpleProductFirst.name$"/>
+        </actionGroup>
+        <actionGroup ref="AdminCheckProductIsMissingInCategoryProductsGridActionGroup" stepKey="dontSeeSecondProduct">
+            <argument name="productName" value="$createSimpleProductSecond.name$"/>
+        </actionGroup>
+
+        <!-- Reset Product Filters -->
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml:65]: Same as above.

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        <actionGroup ref="AdminCategoryResetProductFilterActionGroup" stepKey="applyProductFilters"/>
+
+        <!-- Check Products are present in Category -->
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminCategoryProductFiltersTest.xml:68]: Same as above.

**Review by lbajsarowicz — APPROVED** [review]
✔️  Just minor elements to be adjusted (comments).
Except then that - I don't see any issues there.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7149](https://jira.corp.magento.com/browse/ENGCOM-7149) has been created to process this Pull Request

**Comment by engcom-Bravo** [comment]
:heavy_check_mark:  QA Passed

Tested together with the related EE PR. The Filterable section is available and all the filters work as expected

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run Semantic Version Checker

**Comment by gabrieldagama** [comment]
Approval ticket created: https://jira.corp.magento.com/browse/MC-39098

**Comment by sivaschenko** [comment]
@magento give me test instance

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run Functional Tests B2B

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento create issue

**Review by engcom-Bravo — CHANGES_REQUESTED** [review]
:x: QA failed

Tested together with the related EE PR

There is an unneeded filterable field without name 
unneeded

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Pull request closed without merging**